### PR TITLE
feat: implement the `handle_batch_open_requests`

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#4026

## What's changed and what's your intention?

1. Add the `handle_batch_open_requests ` method for the `RegionEngine` trait
2. Implement the `handle_batch_open_requests` method for `mito`. It tries to group opening regions by topic and then open regions with the same topic in a batch.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/store-api/src/region_engine.rs [hunk 1]
```diff
@@ -177,11 +179,38 @@ pub trait RegionScanner: Debug + DisplayAs + Send + Sync {
...
     fn name(&self) -> &str;
 
+    /// Handles batch open region requests.
+    async fn handle_batch_open_requests(
+        &self,
+        parallelism: usize,
+        requests: Vec<(RegionId, RegionOpenRequest)>,
+    ) -> Result<BatchResponses, BoxedError> {
+        let semaphore = Arc::new(Semaphore::new(parallelism));
+        let mut tasks = Vec::with_capacity(parallelism);
```
Comment (evenyag) [inline src/store-api/src/region_engine.rs:196]:
```suggestion
        let mut tasks = Vec::with_capacity(requests.len());
```

File: src/mito2/src/worker/handle_open.rs [hunk 2]
```diff
@@ -103,6 +105,9 @@ impl<S: LogStore> RegionWorkerLoop<S> {
                 return;
             }
         };
+        if let Some(wal_entry_receiver) = wal_entry_receiver {
+            opener = opener.wal_entry_reader(Box::new(wal_entry_receiver));
+        };
```
Comment (evenyag) [inline src/mito2/src/worker/handle_open.rs:110]: Let the opener take an optional reader.
```suggestion
        }
        .wal_entry_reader(wal_entry_receiver.map(|receiver| Box::new(receiver) as _));
```

File: src/mito2/src/test_util.rs [hunk 3]
```diff
@@ -107,7 +107,7 @@ pub(crate) fn kafka_log_store_factory() -> Option<LogStoreFactory> {
 #[rstest]
 #[case::with_raft_engine(raft_engine_log_store_factory())]
 #[case::with_kafka(kafka_log_store_factory())]
-#[tokio::test]
+#[tokio::test(flavor = "multi_thread")]
```
Comment (evenyag) [inline src/mito2/src/test_util.rs:110]: Why do we need this?

File: src/mito2/src/wal/entry_distributor.rs [hunk 4]
```diff
@@ -114,19 +114,22 @@ impl WalEntryReceiver {
...
-        let WalEntryReceiver {
-            region_id: expected_region_id,
-            mut entry_receiver,
-            arg_sender,
-        } = self;
+    fn read(&mut self, provider: &Provider, start_id: EntryId) -> Result<WalEntryStream<'static>> {
+        let mut arg_sender = self
+            .arg_sender
+            .take()
+            .context(error::InvalidWalReadRequestSnafu {
```
Comment (evenyag) [inline src/mito2/src/wal/entry_distributor.rs:128]:
```suggestion
            .with_context(|| error::InvalidWalReadRequestSnafu {
```

File: src/mito2/src/engine/batch_open_test.rs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+            let request = CreateRequestBuilder::new()
+                .region_dir(&region_dir(region_id))
+                .kafka_topic(topic.clone())
+                .build();
+            let column_schemas = rows_schema(&request);
+            engine
+                .handle_request(region_id, RegionRequest::Create(request))
+                .await
+                .unwrap();
+            barrier.wait().await;
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/mito2/src/engine/batch_open_test.rs:69]: Let the writing happen at the same time

File: src/mito2/src/engine.rs [hunk 6]
```diff
@@ -244,6 +297,106 @@ impl EngineInner {
...
+            let mut tasks = Vec::with_capacity(remaining_region_requests.len());
+            for (region_id, request) in remaining_region_requests {
+                tasks.push((region_id, async move {
+                    let (request, receiver) =
+                        WorkerRequest::new_open_region_request(region_id, request, None);
+
+                    self.workers.submit_to_worker(region_id, request).await?;
+
+                    receiver.await.context(RecvSnafu)?
+                }))
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/mito2/src/engine.rs]: 🥲Added

File: src/mito2/src/engine.rs [hunk 7]
```diff
@@ -244,6 +297,106 @@ impl EngineInner {
...
+        if !remaining_region_requests.is_empty() {
+            let mut tasks = Vec::with_capacity(remaining_region_requests.len());
+            for (region_id, request) in remaining_region_requests {
+                tasks.push((region_id, async move {
+                    let (request, receiver) =
+                        WorkerRequest::new_open_region_request(region_id, request, None);
+
+                    self.workers.submit_to_worker(region_id, request).await?;
+
+                    receiver.await.context(RecvSnafu)?
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/mito2/src/engine.rs:378]: I need the results like `(region_id, Result<rows>)`, maybe store the `region_ids` and `results` separately?

File: src/mito2/src/engine/batch_open_test.rs [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+    let results = engine
+        .handle_batch_open_requests(3, requests)
+        .await
+        .unwrap();
+    for (_, result) in results {
+        assert_eq!(
+            result.unwrap_err().status_code(),
+            StatusCode::RegionNotFound
+        );
+    }
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/mito2/src/engine/batch_open_test.rs:205]: > What if only a subset of regions are absent? We might also test partial failure.

It's covered in `test_batch_open`.
```rust
let mut results = engine
        .handle_batch_open_requests(4, requests)
        .await
        .unwrap();
    let (_, result) = results.pop().unwrap();
    assert_eq!(
        result.unwrap_err().status_code(),
        StatusCode::RegionNotFound
    );
```

File: src/mito2/src/engine/batch_open_test.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/engine/batch_open_test.rs:69]: Why not do it in a for loop
```rust
for row in rows {
    for region in regions {
        write_region(row);
    }
}
```

**Review by evenyag — APPROVED** [review]
LGTM

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**